### DISCUSSION TITLE
Avoiding complications from JEP 472

### DISCUSSION THREAD

**Discussion by Legioth** [oshi/oshi#2698] [Q&A] [answered]
[JEP 472](https://openjdk.org/jeps/472) which is targeted for Java 24 will make Java log warnings whenever custom native code is loaded and some future Java version will go further by blocking custom native code unless explicitly allowed by launching the JVM with the `--enable-native-access` flag enabled.

I assume this will also impact OSHI through the way it relies on JNA. This will in turn mean that any application that uses OSHI will have to make their users' lives slightly more complicated. Is there any possibility of avoiding this in the future? I'm thinking of e.g. offering a limited subset of functionality that relies on launching separate processes using binaries included in the OS to find the same information that is now accessed directly through running native code directly from the JVM process. What are the possibilities here?

**Comment by dbwiddis** [comment]
I've spent a lot of time trying to figure out how to incorporate new FFM features while still supporting older JVMs (I've committed to keeping JDK 8 compatibility through EOL).

I started https://github.com/oshi/oshi-ffm with a vision toward building a new FFM-focused version with limited and modular capabilities.  Of course, I could only really get started with JDK 19 and even with JDK 21 release the whole JEP 454 implementation was incomplete.  

So I've been waiting for the next major release to target some sort of solution, debating between existing oshi-core with some files swapped out in the build (similar to how we add the module descriptor for the java11 build) but that would still be a hybrid maintaining `oshi-core` compatibility and just leveraging FFM for performance improvments where possible.

Your request sounds more like my original vision for `oshi-ffm`.

I'd welcome collaboration on that project; heck, I'd even welcome collaboration on developing a whole new JNA equivalent using only FFM.  It's been discussed on the JNA mailing list.  But the emphasis here is on "collaboration".  I had the issue announcing oshi-ffm pinned for 2 years and got only one committer interested in contributing. I certainly don't have the time or even expertise to do it myself, and don't want to.

I think the community needs this, but this needs a community effort.

So +1 on starting from scratch with a new smaller FFM-only API.  I'm all for it.  I'll help contribute as part of a team.  I can't/won't write it all myself. :)

**Reply by Legioth** [reply]
Does FFM really help in this case? The JEP mentions that e.g. `SymbolLookup::libraryLookup` and `Linker::downcallHandle` also have similar restrictions. Seems like `oshi-ffm` currently uses at least `downcallHandle`.

**Reply by dbwiddis** [reply]
I got started on JDK 19 in preview and used what was available.  

I really only got one class mostly migrated, enough to get some informal benchmarks to know whether it was even worth it.

Obviously things have changed. :)

I wouldn't mind wiping everything and starting from scratch.  Or starting it elsewhere.  

I'm all for this idea.  Let me know how I can participate in something that is community-driven. :)

**Comment by dbwiddis** [comment, marked as answer]
@Legioth with https://github.com/oshi/oshi/pull/3238 you can now just depend on oshi-common and enjoy most OSHI features without `--enable-native-access`.

```java
SystemInfoProvider si = SystemInfoFactory.create();
HardwareAbstractionLayer hal = si.getHardware();
OperatingSystem os = si.getOperatingSystem();
```
